README.md
<img src="http://www.omniosce.org/OmniOSce_logo.svg" height="128">

pkgmgr
=========

[![Build Status](https://travis-ci.org/omniosorg/pkgmgr.svg?branch=master)](https://travis-ci.org/omniosorg/pkgmgr)

Version: 0.3.1

Date: 2019-06-09

pkgmgr is an IPS package management and publishing tool.

Setup
-----

To build `pkgmgr` you require perl and gcc packages on your
system.

Get a copy of `pkgmgr` from https://github.com/omniosorg/pkgmgr/releases
and unpack it into your scratch directory and cd there.

    ./configure --prefix=$HOME/opt/pkgmgr
    make

Configure will check if all requirements are met and give
hints on how to fix the situation if something is missing.

Any missing perl modules will be built and installed into the prefix
directory. Your system perl will NOT be affected by this.

To install the application, just run

    make install

Configuration
-------------

Take a look at the `pkgmgr.conf.dist` file for inspiration.


CHANGES
commit dfa5f2cbb1ff873df7a4b8d228c69449b079b2c0
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sat Jun 8 10:04:18 2019 +0000

    use https for cpanmin.us

commit 8780a70ea861682b9b2674fd2aa1af8fa162ea26
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Thu Jun 6 20:31:03 2019 +0000

    adding staging support for pull command

commit 39b1f22b5493b767825d22ff9b44089c71937a1b
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sat Mar 9 18:39:04 2019 +0100

    ready for 0.3.0

commit 3464a859b02ed7c7fb88a380e37dcff7fbd7541e
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sat Mar 9 18:30:27 2019 +0100

    support repositories that require authentication

commit b7cc6c246eb814a860de957af5cc2357dffcb7f6
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sat Mar 9 17:55:59 2019 +0100

    check certificate expiration date

commit 0307d7a65afc0a99e34a8154d8efc2aa34b9cb3f
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Sep 17 15:42:23 2018 +0200

    ready for 0.2.8

commit 60abde2c97cd4c69b0b42553e6b2c9d28d6725d6
Author: Andy Fiddaman <omnios@citrus-it.co.uk>
Date:   Sun Sep 16 09:53:17 2018 +0000

    Do not check uname if exporting

commit d601f368eaef060420e1f33801832e76b761fa05
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Sep 14 21:58:17 2018 +0200

    ready for 0.2.7

commit b75aec22d6b2067e07c3be846bbf885d386e6339
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Sep 14 20:27:59 2018 +0200

    fix to handle 'reverse components of dash-revision'

commit 905166a107bff7d5240e91494422bd24ff248600
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed May 30 19:05:51 2018 +0200

    don't use --dcert and --dkey for local signing

commit 6fb0269810b3fcc2ebf85132a2d713096e623970
Author: Andy Fiddaman <omnios@citrus-it.co.uk>
Date:   Mon Apr 30 15:38:48 2018 +0000

    pkgmgr diff should treat obsolete/renamed packages as gone

commit 31a905d8baa8d94d9a82fc8deaf892d8dc37c29d
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Apr 25 18:34:45 2018 +0200

    url option added

commit c184db46c685bc4b2d89d79d4a518fe6eacefd2f
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Apr 4 16:51:33 2018 +0200

    ready for 0.2.6

commit d21174c80e0180772a7a7298b5c3f51b98b1cc44
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Tue Apr 3 17:42:16 2018 +0200

    no need for grep

commit c56f7e7c349c640e67a1f669cf5b1351255f95fc
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Tue Apr 3 17:01:15 2018 +0200

    'sign' option added

commit 6e4cc15008b89c973a36c739fd69af73ae9809fa
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Mar 5 13:12:22 2018 +0100

    do not rebuild repository on dry-runs

commit 5ee93a905126b71266289aeaa81b8ed6ab9ac102
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Mar 2 00:41:02 2018 +0100

    relax regexp to handle branches containing hyphens

commit 033a3e740e3132e2a2b74811623a39dcaeb3d082
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Feb 28 16:15:18 2018 +0100

    ready for 0.2.5

commit 11dd52743cc6a73bf0e7d472afd91c4304baffdf
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Feb 28 15:15:42 2018 +0100

    add uname check

commit fbbc2fb6815fc0c62bf4cf94a8b4c22926188b47
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Feb 19 23:02:03 2018 +0100

    ready for 0.2.4

commit 199b5a8945c9def57a0b0bb37e1856fc7883fdf7
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Feb 19 22:14:24 2018 +0100

    add 'auto_rebuild'

commit b41b62905c259f1c66886149f1acff5dc23126f5
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Jan 12 20:15:31 2018 +0100

    fix regression for 'diff' caused by dash versions

commit 1da61e18df9ad2fcf050e3194d214ae095606b8c
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Jan 10 13:23:57 2018 +0100

    ready for 0.2.3

commit 8251ea054b882def1517f97fa01b97fc989abad4
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Jan 10 12:31:45 2018 +0100

    support dash-revisions

commit 7dc115cde2da8e9b942f27da7a90ee0d8ca2a6f4
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Nov 10 15:50:23 2017 +0100

    also test perl 5.26

commit 7d1e7bf38c7ca8c274e6b2476bb32f12c6aa11d7
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Wed Nov 1 11:56:54 2017 +0100

    ready for 0.2.2

commit f002791c3b64188a4b77a3b42a92ea6dc09baaa8
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Oct 30 17:34:34 2017 +0100

    sign packages in one transaction

commit 3d2305fd0fb48a5b3b11fe65cfd891b2be62d0cb
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Tue Oct 24 20:47:00 2017 +0200

    diff option enhanced

commit 4bf6eb740b2ecd9f49cf4119a458fb57d56150d3
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Fri Oct 13 14:39:01 2017 +0200

    ready for 0.2.1

commit 845f62f8a406b766a559e31c1ed02354f0c022e9
Author: Andy Fiddaman <omnios@citrus-it.co.uk>
Date:   Fri Oct 13 11:09:24 2017 +0000

    Produce diffs in markdown format

commit 41ec167a9237bdd5896ad56db78600ac82d3aade
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Tue Oct 10 10:56:24 2017 +0200

    fix config template to match ooce signing policy

commit 007f51f811e226a0ca56e4076668ecbbaac23509
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Sep 25 17:10:06 2017 +0200

    diff: show full pkg name

commit 7b1daedac562153f1cb577e821cf9339e6bcb349
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Thu Sep 21 14:59:18 2017 +0200

    ready for 0.2.0

commit df85f22ae03198850b350697f6e662d97e139100
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Thu Sep 21 13:54:27 2017 +0200

    pull from user defined sources

commit e8f986a264d35bcaeaa7d68db9906e4d995eed6c
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Mon Sep 18 23:33:22 2017 +0200

    cleanup

commit 875578a5372f85ba506fc43c24c6d7c8ff4f1d02
Author: Dominik Hassler <hadfl@omniosce.org>
Date:   Sat Sep 16 17:56:05 2017 +0200

    ready for 0.1.12

commit 2c9a0e987dfcd8d708a48cd8afecddebf5f77636

lib/PkgMgr/Utils.pm
package PkgMgr::Utils;

use strict;
use warnings;

use POSIX qw(isatty);
use Crypt::OpenSSL::X509;
use Time::Seconds qw(ONE_MONTH);

my @RSYNC = qw(/usr/bin/rsync -ahh --stats --delete-after);

# constructor
sub new {
    my $class = shift;
    my $self  = { @_ };

    return bless $self, $class
}

# public methods
sub getNameVersion {
    my $self = shift;
    return (shift =~ m|^pkg://[^/]+/([^@]+)\@([^,]+)|);
}

sub file {
    my $self = shift;
    my $op   = shift;
    my $msg  = shift;

    return sub {
        my $file = shift;
        return open (my $fh, $op, $file) ? undef : "$msg '$file': $!";
    }
}

sub regexp {
    my $self = shift;
    my $rx   = shift;
    my $msg  = shift;

    return sub {
        my $value = shift;
        return $value =~ /$rx/ ? undef : "$msg ($value)";
    }
}

sub elemOf {
    my $self = shift;
    my $elems = [ @_ ];

    return sub {
        my $value = shift;
        return (grep { $_ eq $value } @$elems) ? undef
            : 'expected a value from the list: ' . join(', ', @$elems);
    }
}

sub x509Cert {
    my $self = shift;

    return sub {
        local $@;
        my $x509 = eval {
            local $SIG{__DIE__};
            Crypt::OpenSSL::X509->new_from_file(shift);
        };
        return $@ if $@;

        print STDERR "\n*** WARNING: your certificate will expire on " . $x509->notAfter . "! ***\n\n"
            if $x509->checkend(ONE_MONTH);

        return undef;
    }
}

sub isaTTY {
    my $self = shift;
    return isatty(*STDIN);
}

sub getSTDIN {
    my $self = shift;
    return $self->isaTTY() ? [] : [ split /[\s\n]+/, do { local $/; <STDIN>; } ];
}

sub prettySize {
    my $self = shift;
    my $size = shift;

    my @units = qw(bytes KiB MiB GiB TiB);
    my $i     = $size <= 0 ? 0 : int (log ($size) / log (1024));

    return sprintf("%.2f %s", $size / 1024 ** $i, $units[$i]);
}

sub rsync {
    my $self = shift;

    my @cmd = (@RSYNC, shift, shift);

    system (@cmd) && die "ERROR: executing 'rsync': $!\n";
}

1;

__END__

=head1 COPYRIGHT

Copyright 2019 OmniOS Community Edition (OmniOSce) Association.

=head1 LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.
This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.
You should have received a copy of the GNU General Public License along with
this program. If not, see L<http://www.gnu.org/licenses/>.

=head1 AUTHOR

S<Andy Fiddaman E<lt>omnios@citrus-it.co.ukE<gt>>
S<Dominik Hassler E<lt>hadfl@omniosce.orgE<gt>>

=head1 HISTORY

2017-09-06 had Initial Version

=cut


lib/PkgMgr/Config.pm
package PkgMgr::Config;

use strict;
use warnings;

use JSON::PP;
use FindBin;
use File::Basename qw(basename);
use PkgMgr::Utils;
use Data::Processor;

# constants
my $CONFILE = "$FindBin::RealBin/../etc/" . basename($0) . '.conf'; # CONFFILE

my $SCHEMA = sub {
    my $sv = PkgMgr::Utils->new();

    return {
    GENERAL => {
        optional => 1,
        members  => {
            cert_file        => {
                description  => 'path to certificate file',
                example      => '"cert_file" : "/omniosorg/ssl/certs/ooce_cert.pem"',
                validator    => $sv->x509Cert,
            },
            key_file         => {
                description  => 'path to certificate key file',
                example      => '"key_file" : "/omniosorg/ssl/private/ooce_key.pem"',
                validator    => $sv->file('<', 'Cannot open file'),
            },
            connect_timeout  => {
                optional     => 1,
                description  => 'Seconds to wait trying to connect during transport operations.',
                example      => '"connect_timeout" : "60"',
                default      => 60,
                validator    => $sv->regexp(qr/^\d+$/, 'not a number'),
            },
            lowspeed_timeout => {
                optional     => 1,
                description  => 'Seconds below the lowspeed limit (1024 bytes/second) during '
                              . 'transport operations before the client aborts the operation.',
                example      => '"lowspeed_timeout" : "30"',
                default      => 30,
                validator    => $sv->regexp(qr/^\d+$/, 'not a number'),
            },
            auto_rebuild     => {
                optional     => 1,
                description  => 'automatically rebuild catalog/index after publishing',
                example      => '"auto_rebuild" : "yes"',
                validator    => $sv->elemOf(qw(yes no)),
            },
        },
    },
    REPOS   => {
        members  => {
            '\S+'   => {
                regex   => 1,
                members => {
                    signing     => {
                        description => 'sign packages (yes/no)',
                        example     => '"signing" : "yes"',
                        validator   => $sv->elemOf(qw(yes no)),
                    },
                    src_repo     => {
                        description => 'source (local) repository',
                        example     => '"src_repo" : "/omniosorg/_r22_repo"',
                        validator   => $sv->regexp(qr|^[-\w/.:_]+$|, 'not a valid repo path/URL'),
                    },
                    dst_repo     => {
                        description => 'destination (remote) repository',
                        example     => '"dst_repo" : "https://pkg.omniosce.org/r151022/core"',
                        validator   => $sv->regexp(qr|^[-\w/.:_]+$|, 'not a valid repo path/URL'),
                    },
                    staging_repo => {
                        optional    => 1,
                        description => 'staging repository',
                        example     => '"staging_repo" : "https://pkg.omniosce.org/r151022/staging"',
                        validator   => $sv->regexp(qr|^[-\w/.:_]+$|, 'not a valid repo path/URL'),
                    },
                    publisher   => {
                        description => 'publisher name',
                        example     => '"publisher" : "omnios"',
                        validator   => $sv->regexp(qr/^[\w.]+$/, 'not a valid publisher name'),
                    },
                    release     => {
                        description => 'release',
                        example     => '"release" : "r151022"',
                        transformer => sub { return (shift =~ /^r?(.*)$/)[0]; }, # remove leading 'r' if given
                        validator   => $sv->regexp(qr/^1510\d{2}$/, 'not a valid release'),
                    },
                    on_hold     => {
                        optional    => 1,
                        array       => 1,
                        description => 'list of FMRI patterns (regexp) not to stage or publish',
                        example     => '"on_hold" : [ "binutils", "cherrypy" ]',
                        validator   => sub { return undef },
                    },
                    pull_src    => {
                        optional    => 1,
                        members     => {
                            '\S+'   => {
                                regex       => 1,
                                description => 'pull source',
                                example     => '"host1" : "user@build-host:/build/repo"',
                                validator   => $sv->regexp(qr/^.*$/, 'expected a string'),
                            },
                        },
                    },
                    restricted => {
                        optional    => 1,
                        description => 'restricted repository; authentication needed (yes/no)',
                        example     => '"restricted" : "no"',
                        default     => 'no',
                        validator   => $sv->elemOf(qw(yes no)),
                    },
                },
            },
        },
    },
    }
};

# constructor
sub new {
    my $class = shift;
    my $self  = { @_ };

    $self->{cfg} = Data::Processor->new($SCHEMA->());

    return bless $self, $class;
}

sub loadConfig {
    my $self     = shift;
    my $repo     = shift;
    my $confFile = shift // $CONFILE;

    open my $fh, '<', $confFile or die "ERROR: opening config file '$confFile': $!\n";
    my $configJSON = do { local $/; <$fh>; };
    close $fh;

    my $config = JSON::PP->new->decode($configJSON);

    my $ec = $self->{cfg}->validate($config);
    $ec->count and die join ("\n", map { $_->stringify } @{$ec->{errors}}) . "\n";

    exists $config->{REPOS}->{$repo}
        or die "ERROR: repository '$repo' not defined in config file.\n";

    return $config;
}

1;

__END__

=head1 COPYRIGHT

Copyright 2019 OmniOS Community Edition (OmniOSce) Association.

=head1 LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.
This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.
You should have received a copy of the GNU General Public License along with
this program. If not, see L<http://www.gnu.org/licenses/>.

=head1 AUTHOR

S<Andy Fiddaman E<lt>omnios@citrus-it.co.ukE<gt>>
S<Dominik Hassler E<lt>hadfl@omniosce.orgE<gt>>

=head1 HISTORY

2017-09-06 had Initial Version

=cut


lib/PkgMgr.pm
package PkgMgr;

use strict;
use warnings;
use Time::Piece;
use IO::Uncompress::Gunzip qw(gunzip $GunzipError);

# constants/tools
my $PKGREPO = '/usr/bin/pkgrepo';
my $PKGRECV = '/usr/bin/pkgrecv';
my $PKGSIGN = '/usr/bin/pkgsign';
my $PKG     = '/usr/bin/pkg';

my %TIME_FACTOR = (
    s   => 1,
    M   => 60,
    h   => 3600,
    d   => 3600 * 24,
    m   => 3600 * 24 * 30,
    y   => 3600 * 24 * 365,
);

# private methods
my $getEpoch = sub {
    return Time::Piece->strptime(shift, '%Y%m%dT%H%M%SZ')->epoch;
};

my $getOptEpoch = sub {
    my $timeOpt = shift;

    # if $timeOpt is an ISO timestamp
    return $getEpoch->($timeOpt) if ($timeOpt =~ /^\d+T\d+Z$/);

    my ($value, $unit) = $timeOpt =~ /^(\d+)(\w)?$/
        or die "ERROR: invalid interval '$timeOpt'.\n";
    # default to seconds
    $unit //= 's';
    exists $TIME_FACTOR{$unit} or die "ERROR: invalid time suffix '$unit'.\n";

    return gmtime->epoch - $value * $TIME_FACTOR{$unit};
};

my $extractPublisher = sub {
    return (shift->{'pkg.fmri'} =~ m|^pkg://([^/]+)|)[0];
};

my $getReleasePublisher = sub {
    my $config = shift;
    my $repo   = shift;

    return ($config->{REPOS}->{$repo}->{release}, $config->{REPOS}->{$repo}->{publisher});
};

# constructor
sub new {
    my $class = shift;
    my $self  = { @_ };

    return bless $self, $class
}

sub hasStaging {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;

    return exists $config->{REPOS}->{$repo}->{staging_repo};
}

sub needsSigning {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;

    return $config->{REPOS}->{$repo}->{signing} eq 'yes';
}

sub isSigned {
    my $self     = shift;
    my $repoPath = shift;
    my $fmri     = shift;

    my @cmd = ($PKG, qw(contents -g), $repoPath, '-m', $fmri);

    open my $pkg, '-|', @cmd or die "ERROR: executing '$PKG'.\n";

    return grep { /^signature/ } (<$pkg>);
}

sub getRepoPath {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opt    = shift;

    $opt->{staging} && do {
        exists $config->{REPOS}->{$repo}->{staging_repo}
            or die "ERROR: no staging repository defined in config file.\n";

        return $config->{REPOS}->{$repo}->{staging_repo};
    };

    return $config->{REPOS}->{$repo}->{$opt->{dst} ? 'dst_repo' : 'src_repo'};
}

sub fetchPackages {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;
    my $fmri   = shift;

    $fmri = [ '*' ] if !@$fmri;
    my $epoch = $opts->{t} ? $getOptEpoch->($opts->{t}) : 0;

    my $repoPath = $self->getRepoPath($config, $repo, $opts);

    my @cert = $config->{REPOS}->{$repo}->{restricted} ne 'yes' ? ()
        : ('--key',  $config->{GENERAL}->{key_file},
           '--cert', $config->{GENERAL}->{cert_file});

    my @cmd = ($PKGREPO, qw(list -F json -s), $repoPath, @cert, @$fmri);
    open my $cmd, '-|', @cmd or die "ERROR: executing '$PKGREPO': $!\n";

    my ($release, $publisher) = $getReleasePublisher->($config, $repo);

    my $packages = [
        grep { $_->{branch} =~ /^(?:$release\.\d+|\d+\.$release)$/
            && $extractPublisher->($_) eq $publisher
            && $getEpoch->($_->{timestamp}) > $epoch
        } @{JSON::PP->new->decode(<$cmd> // '[]')}
    ];

    if ($opts->{long}) {
        for my $p (@$packages) {
            $p->{$_}     = 0 for qw(size csize files);
            $p->{signed} = $self->isSigned($repoPath, $p->{'pkg.fmri'});

            my @cmd = ($PKG, qw(contents -H -g), $repoPath,
                '-o', 'value,pkg.size,pkg.csize', $p->{'pkg.fmri'});

            open my $pkg, '-|', @cmd
                or die "ERROR: executing '$PKG'.\n";

            while (<$pkg>) {
                my ($size, $csize) = /^\s*(\d+)\s+(\d+)/ or next;

                $p->{files}++;
                $p->{size}  += $size;
                $p->{csize} += $csize;
            }
            close $pkg;
        }
    }

    return $packages;
}

sub signPackages {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;
    my $pkgs   = shift;

    my $repoPath = $self->getRepoPath($config, $repo, $opts);

    my @cert = $opts->{src} ? ()
        : ('--dkey',  $config->{GENERAL}->{key_file},
           '--dcert', $config->{GENERAL}->{cert_file});

    my @cmd = ($PKGSIGN, '-c', $config->{GENERAL}->{cert_file},
        '-k', $config->{GENERAL}->{key_file},
        ($opts->{n} ? '-n' : ()), '-s', $repoPath, @cert, @$pkgs);

    system (@cmd) && die "ERROR: signing packages: $!\n";
}

sub getSrc {
    my $self    = shift;
    my $config  = shift;
    my $repo    = shift;
    my $opts    = shift;

    return ($opts->{pull} && !$opts->{staging})
           || (($opts->{export} || $opts->{sign}) && $opts->{dst})     ? { dst     => 1 }
         : $opts->{pull}
           || ($opts->{publish} && $self->hasStaging($config, $repo))
           || (($opts->{export} || $opts->{sign}) && $opts->{staging}) ? { staging => 1 }
         :                                                               { src     => 1 }
}

sub getSrcDstRepos {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;

    return ($self->getRepoPath($config, $repo, $opts), $opts->{d})
        if $opts->{export};

    my $srcRepo = $self->getRepoPath($config, $repo,
                  $opts->{pull} && !$opts->{staging}      ? { dst     => 1 }
                : $self->hasStaging($config, $repo)
                    && (!$opts->{stage} || $opts->{pull}) ? { staging => 1 }
                :                                           { src     => 1 }
    );

    my $dstRepo = $self->getRepoPath($config, $repo,
                  $opts->{pull}  ? { src     => 1 }
                : $opts->{stage} ? { staging => 1 }
                :                  { dst     => 1 }
    );

    return ($srcRepo, $dstRepo);
}

sub filterOnHold {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $pkgs   = shift;

    # nothing 'on hold'. we are done.
    return $pkgs if !exists $config->{REPOS}->{$repo}->{on_hold};

    # build a package list excluding packages that match an 'on hold' pattern
    my @pkgs = ();
    PKG: for my $pkg (@$pkgs) {
        $pkg =~ /$_/ && next PKG for @{$config->{REPOS}->{$repo}->{on_hold}};
        push @pkgs, $pkg;
    }

    return \@pkgs;
}

sub publishPackages {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;
    my $pkgs   = shift;

    my ($srcRepo, $dstRepo) = $self->getSrcDstRepos($config, $repo, $opts);

    my @cert = $opts->{pull} || $opts->{export} ? ()
        : ('--dkey',  $config->{GENERAL}->{key_file},
           '--dcert', $config->{GENERAL}->{cert_file});

    push @cert, $config->{REPOS}->{$repo}->{restricted} ne 'yes' ? ()
        : ('--key',  $config->{GENERAL}->{key_file},
           '--cert', $config->{GENERAL}->{cert_file});

    # set timeout env variables
    $ENV{PKG_CLIENT_CONNECT_TIMEOUT}  = $config->{GENERAL}->{connect_timeout};
    $ENV{PKG_CLIENT_LOWSPEED_TIMEOUT} = $config->{GENERAL}->{lowspeed_timeout};

    my @cmd = ($PKGRECV, ($opts->{n} ? '-n' : ()), ($opts->{export} ? '-a' : ()),
        '-s', $srcRepo, '-d', $dstRepo, @cert, qw(-m latest), @$pkgs);

    system (@cmd) && die 'ERROR: ' . ($opts->{pull} ? 'pulling' : 'publishing') . " packages: $!\n";
}

sub removePackages {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;
    my $pkgs   = shift;

    my $repoPath = $self->getRepoPath($config, $repo, { src => 1 });

    my @cmd = ($PKGREPO, qw(remove -s), $repoPath, ($opts->{n} ? '-n' : ()), @$pkgs);

    system (@cmd) && die "ERROR: cannot remove packages from repo '$repoPath'.\n";
}

sub rebuildRepo {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;

    my $repoPath = $self->getRepoPath($config, $repo, $opts);

    my @cmd = ($PKGREPO, qw(rebuild -s), $repoPath, ($opts->{staging} || $opts->{dst} ? ('--key',
        $config->{GENERAL}->{key_file}, '--cert', $config->{GENERAL}->{cert_file}) : ()));

    system (@cmd) && die "ERROR: rebuilding repo '$repoPath'.\n";
}

sub checkUname {
    my $self   = shift;
    my $config = shift;
    my $repo   = shift;
    my $opts   = shift;
    my $pkgs   = shift;

    my ($srcRepo, $dstRepo)   = $self->getSrcDstRepos($config, $repo, $opts);
    my ($release, $publisher) = $getReleasePublisher->($config, $repo);
    my $branch = $release =~ /[13579]$/ ? 'master' : "r$release";

    my ($pkg, $ver) =
        map { local $_ = $_; s/([^^A-Za-z0-9\-_.!~*'()])/ sprintf "%%%0X", ord $1 /eg; $_ }
        map { local $_ = $_; s|^pkg://$publisher/||; split /\@/, $_, 2 }
        grep { m|system/kernel/platform| } @$pkgs or return;

    # only checking if publishing from local repository
    return if $srcRepo =~ /^http/;

    print "Checking uname...\n\n";

    open my $fh, '<', "$srcRepo/publisher/$publisher/pkg/$pkg/$ver"
        or die "ERROR: cannot open manifest: $!\n";

    my ($hash, $pref) = map {
        m|^\S+\s+((\S{2})\S+).*path=platform/i86pc/kernel/amd64/unix.*debug\.illumos=false|
            ? ($1, $2) : ()
    } (<$fh>)
        or die "ERROR: hash file missing\n";

    close $fh;

    my $contents;
    gunzip "$srcRepo/publisher/$publisher/file/$pref/$hash" => \$contents
        or die "ERROR: gunzip hash file failed: $GunzipError\n";

    my ($uname) = $contents =~ /(omnios-.+?-[\da-f]{10})/
        or die "ERROR: cannot extract uname.\n";

    $uname =~ /^omnios-$branch/
        or die "ERROR: publishing from wrong branch: $uname\n";
}

1;

__END__

=head1 COPYRIGHT

Copyright 2019 OmniOS Community Edition (OmniOSce) Association.

=head1 LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.
This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.
You should have received a copy of the GNU General Public License along with
this program. If not, see L<http://www.gnu.org/licenses/>.

=head1 AUTHOR

S<Andy Fiddaman E<lt>omnios@citrus-it.co.ukE<gt>>
S<Dominik Hassler E<lt>hadfl@omniosce.orgE<gt>>

=head1 HISTORY

2017-09-06 had Initial Version

=cut
